### PR TITLE
Replace vfsgen with embed for static web assets

### PR DESCRIPTION
**Pull request by jan--f**
This replaces #9380. 

Note that this still contains a go mod replace for `prometheus/common`. I'll remove this once https://github.com/prometheus/common/pull/343 has landed.

### CODE REVIEW

File: Makefile.common [hunk 1]
```diff
@@ -238,8 +238,12 @@ endif
 
 .PHONY: common-build
 common-build: promu
+	@echo ">> compressing assets"
+	@scripts/compress_assets.sh
 	@echo ">> building binaries"
 	GO111MODULE=$(GO111MODULE) $(PROMU) build --prefix $(PREFIX) $(PROMU_BINARIES)
```
Comment (juliusv) [inline Makefile.common:244]: If anything goes wrong in this step, the assets stay compressed... could we decompress them not in-place, but to some alternative temporary place, and then embed them from there?

Comment (jan--f) [reply Makefile.common:244]: `embed` will only accept directives that are below or at the same level as the including file (directive may not start the `/` or include `..`) so anything like that would have to happen in `web/ui/`. I think this requirement will mess with builds not built with `builtinassets`.
I figured the assets staying compressed is the least bad solution as git would easily be able to revert that.

Comment (jan--f) [reply Makefile.common]: @juliusv How about this solution: the three steps are executed in a single sub-shell, i.e. if the build has an error, the decompress step is run regardless.

Comment (jan--f) [reply Makefile.common]: I added the cleanup step as a trap on EXIT. This should now always decompress the assets whether the build fails or succeeds, while returning the correct exit code.
@juliusv I'd much appreciate if you could take another look.

**Comment by juliusv** [comment]
Thanks :+1: I think it's good enough for now. I'm still not super happy with introducing that much new "shell noise" into the Makefile in the long term... but I guess the native Go embedding will never support compression of assets?

**Comment by jan--f** [comment]
Thanks, I certainly agree about the shell noise, I tried to keep it to a minimum.
`embed` might well gain better compression support in the future. Its specifically called out in the [design proposal](https://go.googlesource.com/proposal/+/master/design/draft-embed.md#compression-to-reduce-binary-sizel). I'll keep an eye on it.

**Comment by juliusv** [comment]
Thanks, that's an interesting read! I hope they add it, as the tradeoffs they talk about would definitely be better with compression in our case :)

**Comment by roidelapluie** [comment]
I have reverted this pull request for two reasons that we will have to address:

- the replace in the go.mod is breaking downstream because replace are not taken into account in dependencies
- the change in makefile.common should belong in Makefile instead

I will make those changes once common is released. thanks!

**Comment by jan--f** [comment]
Thanks and apologies for the churn.

**Pull request merged**